### PR TITLE
[dataclass_transform] minimal implementation of dataclass_transform

### PR DESCRIPTION
**Pull request by wesleywright**
This is a very simple first step to implementing [PEP 0681](https://peps.python.org/pep-0681/#decorator-function-example), which will allow MyPy to recognize user-defined types that behave similarly to dataclasses.

This initial implementation is very limited: we only support decorator-style use of `typing.dataclass_transform` and do not support passing additional options to the transform (such as `freeze` or `init`).

Within MyPy, we add a new `is_dataclass_transform` field to `FuncBase` which is populated during semantic analysis. When we check for plugin hooks later, we add new special cases to use the existing dataclasses plugin if a class decorator is marked with `is_dataclass_transform`. Ideally we would use a proper plugin API; the hacky special case here can be replaced in subsequent iterations.

### CODE REVIEW

File: mypy/plugins/common.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
     Var,
 )
 from mypy.plugin import CheckerPluginInterface, ClassDefContext, SemanticAnalyzerPluginInterface
-from mypy.semanal import ALLOW_INCOMPATIBLE_OVERRIDE, set_callable_name
+from mypy.semanal_shared import ALLOW_INCOMPATIBLE_OVERRIDE, set_callable_name
```
Comment (wesleywright) [inline mypy/plugins/common.py:22]: this move was needed to avoid a cyclic dependency between `semanal.py` and `plugins/common.py`

**Comment by thomkeh** [comment]
I see only `typing.dataclass_transform` in the code; does this also work with `typing_extensions.dataclass_transform`?

**Comment by wesleywright** [comment]
> I see only `typing.dataclass_transform` in the code; does this also work with `typing_extensions.dataclass_transform`?

I think this is a gap currently, thanks for the catch (this is my first PR adding this kind of feature haha)

File: test-data/unit/fixtures/dataclasses.pyi [hunk 2]
```diff
@@ -37,7 +37,11 @@ class dict(Mapping[KT, VT]):
     def get(self, k: KT, default: Union[KT, _T]) -> Union[VT, _T]: pass
     def __len__(self) -> int: ...
 
-class list(Generic[_T], Sequence[_T]): pass
+class list(Generic[_T], Sequence[_T]):
+    def __contains__(self, item: object) -> int: pass
+    def __getitem__(self, key: int) -> _T: pass
+    def __iter__(self) -> Iterator[_T]: pass
+
```
Comment (wesleywright) [inline test-data/unit/fixtures/dataclasses.pyi:44]: needed to use `dataclasses.pyi` and `typing-full.pyi` at the same time in a test case

**Review by JukkaL — APPROVED** [review]
Thanks for the PR! This is solid basic implementation. You'll need to fix flake8, otherwise everything looks good.

**Review by hauntsaninja — APPROVED** [review]
Thanks, this looks great. Let's get this into 1.0. It's not complete, but it will result in strictly fewer errors for users of PEP 681 than current master.

**Pull request merged**